ddtrace/profiling/collector/_memalloc_tb.cpp

```diff
@@ -61,108 +44,397 @@ traceback_t::init()
+            }
+            Py_DECREF(name_bytes);
+        } else {
+            PyErr_Clear();
+        }
     } else {
-        name = code->co_name ? code->co_name : unknown_name;
-        filename = code->co_filename ? code->co_filename : unknown_name;
+        Py_XDECREF(name_obj);
     }
 
-    Py_INCREF(name);
-    Py_INCREF(filename);
+    // Get thread.native_id attribute
+    PyObject* native_id_obj = PyObject_GetAttrString(thread, "native_id");
+    if (native_id_obj != NULL) {
+        if (PyLong_Check(native_id_obj)) {
+            thread_native_id = PyLong_AsLongLong(native_id_obj);
+        }
+        Py_DECREF(native_id_obj);
+    } else {
+        PyErr_Clear();
+    }
 
-#ifdef _PY39_AND_LATER
-    Py_XDECREF(code);
-#endif
+    Py_DECREF(thread);
+
+    // Push threadinfo to sample with all data
+    sample.push_threadinfo(thread_id, thread_native_id, thread_name);
+
+    // Error will be restored automatically by error_restorer destructor
 }
 
-traceback_t::traceback_t(void* ptr,
-                         size_t size,
-                         PyMemAllocatorDomain domain,
-                         size_t weighted_size,
-                         PyFrameObject* pyframe,
-                         uint16_t max_nframe)
-  : ptr(ptr)
-  , size(weighted_size)
-  , domain(domain)
-  , thread_id(PyThread_get_thread_ident())
-  , reported(false)
-  , count(0)
+/* Helper function to extract code object and line number from a PyFrameObject */
+static void
+extract_frame_info_from_pyframe(PyFrameObject* frame, PyCodeObject** code_out, int* lineno_out)
 {
-    // Size 0 allocations are legal and we can hypothetically sample them,
-    // e.g. if an allocation during sampling pushes us over the next sampling threshold,
-    // but we can't sample it, so we sample the next allocation which happens to be 0
-    // bytes. Defensively make sure size isn't 0.
-    size_t adjusted_size = size > 0 ? size : 1;
-    double scaled_count = ((double)weighted_size) / ((double)adjusted_size);
-    count = (size_t)scaled_count;
+    *code_out = NULL;
+    *lineno_out = 0;
 
-    // Collect frames from the Python frame chain
-    size_t total_nframe = 0;
-    for (; pyframe != NULL;) {
-        if (frames.size() < max_nframe) {
-            frames.emplace_back(pyframe);
-        }
-        // TODO(dsn): add a truncated frame to the traceback if we exceed the max_nframe
-        total_nframe++;
+    int lineno_val = PyFrame_GetLineNumber(frame);
+    if (lineno_val < 0)
+        lineno_val = 0;
 
 #ifdef _PY39_AND_LATER
-        PyFrameObject* back = PyFrame_GetBack(pyframe);
-        Py_DECREF(pyframe);
-        pyframe = back;
+    PyCodeObject* code = PyFrame_GetCode(frame);
 #else
-        pyframe = pyframe->f_back;
+    PyCodeObject* code = frame->f_code;
 #endif
-        memalloc_debug_gil_release();
+
+    *code_out = code;
+    *lineno_out = lineno_val;
+}
+
+#if defined(_PY311_AND_LATER) && !defined(_PY313_AND_LATER)
```

Comment:
@P403n1x87 what can we do in more recent versions? They seem to have removed this from the headers entirely